### PR TITLE
TPP-00063 Onboarding Service AsyncStorage Endpoints

### LINKED ISSUE #63: TPP-00063: Onboarding Service AsyncStorage Endpoints
> As a user...I want the inputs I'm providing the front-end to be saved in a backend, and I want to read from those saved values. 

API documentation: https://github.com/uoftblueprint/the-period-purse/wiki/Onboarding,-Cycle,-Settings-API

In a file titled OnboardingService.js, implement the APIs specified in the API documentation that Onboarding will need. You can check that these are actually being stored/check what's in AsyncStorage by installing the React Native Debugger, instructions found in the README.  

Example of an API Service file Faye made: https://github.com/uoftblueprint/the-period-purse/wiki/Example-API-Service-File

Official AsyncStorage docs: https://react-native-async-storage.github.io/async-storage/docs/usage

for .catch(error) just do console.log(error); for now! We'll have a story go through these later. 

You should be importing it like this: 
```import AsyncStorage from '@react-native-async-storage/async-storage';```
DO NOT USE 
```import AsyncStorage from react-native;```

### PR DESCRIPTION
**Pull request by helen-li**
# Description

* Implemented the Onboarding API specified in the API documentation in wiki in the file OnboardingService.js under the service folder
* Started connecting frontend to backend (the period length page is done here, will continue this for the other pages in my next ticket)

Fixes #63 

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (if required)
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tpp-app/src/onboarding/PeriodStart.js [hunk 1]
```diff
@@ -22,7 +23,11 @@ export default function PeriodStart ({ navigation }) {
 
       <TwoButtonContainer>
         <SkipButton title="Skip" onPress={() => navigation.navigate(STACK_SCREENS["Symptoms Choices"])}/>
-        <NextButton title="Next" onPress={() => navigation.navigate(STACK_SCREENS["Symptoms Choices"])}/>
+        <NextButton title="Next" onPress={() => 
+          {
+            OnboardingService.PostInitialPeriodStart(new Date(2011, 11, 30));
```
Comment (yuemily24) [inline tpp-app/src/onboarding/PeriodStart.js:28]: Is there a reason why this date was chosen?

File: tpp-app/src/services/OnboardingService.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                periodLength = JSON.stringify(periodLength)
+                const initialRes = await AsyncStorage.setItem(KEYS.INITIAL_PERIOD_LENGTH, periodLength);
+                const averageRes = AsyncStorage.setItem(KEYS.AVERAGE_PERIOD_LENGTH, periodLength);
+
+                console.log('Initialized initialPeriodLength and averagePeriodLength as ' + periodLength);
+                return Promise.all([initialRes, averageRes]);
+            }
+        } catch (e) {
+            console.log('There was an error with POSTInitialPeriodLength');
+            console.log(e);
```
Comment (yuemily24) [inline tpp-app/src/services/OnboardingService.js]: If the return type for this function is always going to be a promise, you should add a reject in the catch block

File: tpp-app/src/services/OnboardingService.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+     * Initializes entries for the user's last period.  
+     * This only happens when the user has given inputs for both their 
+     * initial period length and the start date of their last period. 
+     * @param periodStart date representing the start date of their last period 
+     * @returns a promise resolving when the set operation is complete 
+     */
+    PostInitialPeriodStart: async function(periodStart) {
+        try {
+            const periodLength = await AsyncStorage.getItem('initialPeriodLength');
+            if(periodLength != null && periodStart != null) {
```
Comment (yuemily24) [inline tpp-app/src/services/OnboardingService.js]: You can shorten this to `(!periodLength && !periodStart)`. Javascript will automatically check if its null, undefined, empty, etc.

File: tpp-app/src/services/OnboardingService.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                    const res = await AsyncStorage.setItem(JSON.stringify(year), JSON.stringify(yearDicts[year]));
+                    allRes.push(res);
+                })
+                console.log("Initialized the first loggings");
+                return Promise.all(allRes);
+            }
+        }
+        catch (e) {
+            console.log('There was an error with POSTInitialPeriodStart');
+            console.log(e);
```
Comment (yuemily24) [inline tpp-app/src/services/OnboardingService.js]: Once again if the return type is a promise, you should add a reject in the catch block to cover all cases in the function.

File: tpp-app/src/services/OnboardingService.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                    yearDicts[date.getFullYear()][date.getMonth()][date.getDate() - 1] = new Symptoms(FLOW_LEVEL.MEDIUM);
+                });
+
+                let allRes = [];
+                yearsSet.forEach(async (year) => {
+                    const res = await AsyncStorage.setItem(JSON.stringify(year), JSON.stringify(yearDicts[year]));
+                    allRes.push(res);
+                })
+                console.log("Initialized the first loggings");
+                return Promise.all(allRes);
```
Comment (yuemily24) [inline tpp-app/src/services/OnboardingService.js]: `Promise.all` is typically used when you have multiple promises to resolve for. Since we technically only have one here, I recommend changing your function to be
```javascript
const PostInitialPeriodStart = async (periodStart) => new Promise( (resolve, reject) => {
  try {
    ...
    resolve('yay it worked');
  } catch {
    ...
    reject();
  }
});
```
This will also handle my comment below about having a `reject` in the catch block.

File: tpp-app/src/services/OnboardingService.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                let moodRes = await AsyncStorage.setItem(TRACK_SYMPTOMS.TRACK_MOOD, JSON.stringify(mood));
+                let sleepRes = await AsyncStorage.setItem(TRACK_SYMPTOMS.TRACK_SLEEP, JSON.stringify(sleep));
+                let crampsRes = await AsyncStorage.setItem(TRACK_SYMPTOMS.TRACK_CRAMPS, JSON.stringify(cramps));
+                let exerciseRes = await AsyncStorage.setItem(TRACK_SYMPTOMS.TRACK_EXERCISE, JSON.stringify(exercise));
+                console.log("Initialized tracking preferences");
+                return Promise.all([flowRes, moodRes, sleepRes, crampsRes, exerciseRes]);
+            }
+        } catch (e) {
+            console.log('There was an error with POSTSymptomsToTrack');
+            console.log(e);
```
Comment (yuemily24) [inline tpp-app/src/services/OnboardingService.js]: reject here !

File: tpp-app/src/services/utils/constants.js [hunk 7]
```diff
@@ -1,18 +1,17 @@
...
+  SPOTTING: "SPOTTING"
 }
 
 export const MOOD_LEVEL = {
   HAPPY: "HAPPY",
   SAD: "SAD",
   NEUTRAL: "NEUTRAL",
   SICK: "SICK",
-  ANGRY: "ANGRY",
-  CUSTOM: "" // empty string
```
Comment (yuemily24) [inline tpp-app/src/services/utils/constants.js:15]: Are we getting rid of custom? Same goes for the custom in `EXERCISE_TYPE`

File: tpp-app/src/onboarding/PeriodStart.js (hunk 1 above)
Comment (yuemily24) [inline tpp-app/src/onboarding/PeriodStart.js]: Is there a reason why this date was chosen?

**Review by yuemily24 — CHANGES_REQUESTED** [review]
The logic is really easy to follow and it's well commented! Just make sure you're covering all your cases for Promises in a try/catch block. I also added some general questions I'm curious about.

File: tpp-app/src/onboarding/PeriodStart.js (hunk 1 above)
Comment (helen-li) [reply tpp-app/src/onboarding/PeriodStart.js]: This was just me testing hehe it will be replaced once I make the button for picking dates on the frontend!

File: tpp-app/src/services/utils/constants.js (hunk 7 above)
Comment (helen-li) [reply tpp-app/src/services/utils/constants.js:15]: Maybe @faye-tan and @yichenesia can help clarify here but I recall hearing something at a coworking about no longer using custom emojis for mood and exercise?

File: tpp-app/src/services/OnboardingService.js (hunk 5 above)
Comment (helen-li) [reply tpp-app/src/services/OnboardingService.js]: I'm not too familiar with Promises so thanks for all of this Emily! I've added rejects to all methods except here I used `Promise.all` to account for the possibility that I have to call `setItem()` more than once in case I need to initialize a 5-day-long period that started on December 30th of some year. Would the best practice then be to still use `Promise.all` or is there something else that I should use?

File: tpp-app/src/services/utils/constants.js (hunk 7 above)
Comment (faye-tan) [reply tpp-app/src/services/utils/constants.js:15]: Wiki has been updated! We are indeed no longer using custom emojis at all, MOOD_LEVEL and EXERCISE should have 9 defined emojis: https://github.com/uoftblueprint/the-period-purse/wiki/Models-and-AsyncStorage-Keys

File: tpp-app/src/services/OnboardingService.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+import { FLOW_LEVEL, TRACK_SYMPTOMS, KEYS } from './utils/constants.js'
+import { initializeEmptyYear } from "./utils/helpers.js"
+import { Symptoms } from './utils/models.js';
+
+const OnboardingService = {
+    /**
+     * Initializes user's initial period length and average period length. 
+     * @param periodLength number representing the period length
+     * @returns a promise resolving when the two set operations are complete 
+     */
```
Comment (faye-tan) [inline tpp-app/src/services/OnboardingService.js]: i am loving these detailed comments 💃

File: tpp-app/src/services/OnboardingService.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+     * Initializes the user's initial tracking preferences for what symptoms to track. 
+     * At least one symptom has to be selected for initialization to happen. 
+     * @param flow boolean representing whether to track flow
+     * @param mood boolean representing whether to track mood 
+     * @param sleep boolean representing whether to track sleep
+     * @param cramps boolean representing whether to track cramps
+     * @param exercise boolean representing whether to track exercise 
+     * @returns a promise resolving when all the set operations are complete 
+     */
+    PostSymptomsToTrack: async function(flow, mood, sleep, cramps, exercise) {
```
Comment (faye-tan) [inline tpp-app/src/services/OnboardingService.js]: using ```AsyncStorage.multiSet()``` might be a better option here, since you are calling one AsyncStorage function instead of it five times 

https://react-native-async-storage.github.io/async-storage/docs/api/#multiset

Comment (helen-li) [reply tpp-app/src/services/OnboardingService.js]: oooh good call i did not realize this existed - thanks faye!

File: tpp-app/src/services/OnboardingService.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import { Symptoms } from './utils/models.js';
+
+/**
+ * Initializes user's initial period length and average period length. 
+ * @param periodLength a number representing the period length
+ * @returns a promise resolving when the multiset operation is complete 
+ */
+export const PostInitialPeriodLength = async (periodLength) => new Promise( async (resolve, reject) => {
+    try {
+        if(periodLength != null) {
```
Comment (yuemily24) [inline tpp-app/src/services/OnboardingService.js]:
```suggestion
        if(periodLength) {
```
Shortens it and expands the range of checks. This will make sure that it's defined in any way that's not `undefined` or `null`. Fun fact, you can use this for strings too and it'll return false if it's an empty string.

**Review by yuemily24 — CHANGES_REQUESTED** [review]
Left one tiny suggestion. Otherwise, looks good to me!

**Review by yuemily24 — APPROVED** [review]
Looks great, thanks Helen!

**Review by edwardhuahan — APPROVED** [review]
Looks good!

File: tpp-app/src/onboarding/PeriodLength.js [hunk 11]
```diff
@@ -1,28 +1,44 @@
...
         How long does your {'\n'} period usually last?
       </TitleText>
       <BodyText>
         This will help us make our {'\n'} reminders more accurate
       </BodyText>
 
+      <InputContainer>
+        <TextInput style={styles.input} 
+        placeholder="Tap to input" 
+        keyboardType="number-pad"
```
Comment (faye-tan) [inline tpp-app/src/onboarding/PeriodLength.js:29]: This can be fixed as part of your other ticket since it's not in the requirements for this one but I did not see a number pad keyboard when clicking on the input
Screen Shot 2022-02-26 at 6 27 43 PM

Comment (helen-li) [reply tpp-app/src/onboarding/PeriodLength.js:29]: ooo yep for me the number pad would only pop up after i enable the keyboard with cmd k i will look more into this

File: tpp-app/src/services/OnboardingService.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+ * @param flow boolean representing whether to track flow
+ * @param mood boolean representing whether to track mood 
+ * @param sleep boolean representing whether to track sleep
+ * @param cramps boolean representing whether to track cramps
+ * @param exercise boolean representing whether to track exercise 
+ * @returns a promise resolving when the multiset operation is complete 
+ */
+export const PostSymptomsToTrack = async (flow, mood, sleep, cramps, exercise) => new Promise( async (resolve, reject) => {
+    try {
+        if([flow, mood, sleep, cramps, exercise].some((bool) => bool)) {
```
Comment (faye-tan) [inline tpp-app/src/services/OnboardingService.js:91]: should .every() be used instead?

Comment (helen-li) [reply tpp-app/src/services/OnboardingService.js:91]: i thought we just require that at least one of the tracking options to be true here so i used some()?

Comment (faye-tan) [reply tpp-app/src/services/OnboardingService.js:91]: oh u right :0 yes!

**Review by faye-tan — APPROVED** [review]
